rx888_source/src/main.cpp

```diff
@@ -76,188 +159,261 @@ class RX888SourceModule : public ModuleManager::Instance {
         }
-
-        delete[] buffer;
-
-        // Stop FX3 chip
-        fx3Control(STOPFX3);
+        ImGui::Text("ADC Dither");
+        ImGui::SameLine();
+        if (ImGui::Checkbox(CONCAT("##_rx888_adc_dither_", _this->name), &_this->adc_dither)) {
+            sddc_set_adc_dither(_this->sddc, _this->adc_dither ? 1 : 0);
+        }
+        ImGui::Text("ADC Random");
+        ImGui::SameLine();
+        if (ImGui::Checkbox(CONCAT("##_rx888_adc_random_", _this->name), &_this->adc_random)) {
+            sddc_set_adc_random(_this->sddc, _this->adc_random ? 1 : 0);
+        }
     }
 
-    static void _worker(RX888SourceModule* _this) {
+#ifdef LIBSDDC_RAW
+    static void asyncHandler(uint32_t len, uint8_t *buf, void *ctx) {
+        RX888SourceModule* _this = (RX888SourceModule*)ctx;
         dsp::complex_t* iqbuffer = new dsp::complex_t[STREAM_BUFFER_SIZE];
+        int blockSize = len/sizeof(float);
+        for (int i = 0; i < blockSize; i++) {
+            iqbuffer[i].re = (float)buf[i*sizeof(float)] / 32768.0f;
+        }
+
+        dsp::complex_t* iqbuffer2 = new dsp::complex_t[STREAM_BUFFER_SIZE];
         lv_32fc_t phase = lv_cmake(1.0f, 0.0f);
 
-        // Read from real stream
-        int count = _this->realStream.read();
-        int blockSize = count;
+        // Calculate the traslation frequency based on the tuning
+        float delta = FL_M_PI / 2.0f;
+        lv_32fc_t phaseDelta = lv_cmake(std::cos(delta), std::sin(delta));
 
-        while (count >= 0) {
-            for (int i = 0; i < count; i++) {
-                iqbuffer[i].q = (float)_this->realStream.readBuf[i] / 32768.0f;
-            }
-            _this->realStream.flush();
-
-            // Calculate the traslation frequency based on the tuning
-            float delta = -(_this->freq / (float)ADC_RATE) * 2.0f * FL_M_PI;
-            lv_32fc_t phaseDelta = lv_cmake(std::cos(delta), std::sin(delta));
-
-            // Apply translation
-            volk_32fc_s32fc_x2_rotator_32fc((lv_32fc_t*)_this->stream.writeBuf, (lv_32fc_t*)iqbuffer, phaseDelta, &phase, count);
-
-            // Decimate
-            blockSize = count;
-            for (int d = 0; d < (_this->decimation - 1); d++) {
-                blockSize = (blockSize >> 1);
-                for (int i = 0; i < blockSize; i++) {
-                    _this->stream.writeBuf[i].i = (_this->stream.writeBuf[i*2].i + _this->stream.writeBuf[(i*2)+1].i) * 0.5f;
-                    _this->stream.writeBuf[i].q = (_this->stream.writeBuf[i*2].q + _this->stream.writeBuf[(i*2)+1].q) * 0.5f;
-                }
-            }
+        // Apply translation
+        volk_32fc_s32fc_x2_rotator_32fc((lv_32fc_t*)iqbuffer2, (lv_32fc_t*)iqbuffer, phaseDelta, &phase, blockSize);
 
-            // Write to output stream
-            _this->stream.swap(blockSize);
+        // Decimate
+        for (int d = 0; d < _this->deciminator+1; d++) {
+            blockSize = (blockSize >> 1);
```

Comment:
This decimation is incorrect. Data has to be filtered to the lowest bandwidth before.

Reply:
I removed this part of code, now it use decimation from libsddc.